cajita/src/Cajita_Hilbert.hpp

```diff
@@ -0,0 +1,838 @@
+{
+    typedef LayoutHilbert3D array_layout;
+
+    // Dimension array
+    size_t dimension[ARRAY_LAYOUT_MAX_RANK];
+
+    enum
+    {
+        is_extent_constructible = true
+    };
+
+    // Defaults
+    LayoutHilbert3D( LayoutHilbert3D const & ) = default;
+    LayoutHilbert3D( LayoutHilbert3D && ) = default;
+    LayoutHilbert3D &operator=( LayoutHilbert3D const & ) = default;
+    LayoutHilbert3D &operator=( LayoutHilbert3D && ) = default;
+
+    /*!
+      \brief Constructor.
+      \param N0 Entries in 1st-Dimension
+      \param N1 Entries in 2nd-Dimension
+      \param N2 Entries in 3rd-Dimension
+      \param N3 Entries in 4th-Dimension
+      \param N4 Entries in 5th-Dimension
+      \param N5 Entries in 6th-Dimension
+      \param N6 Entries in 7th-Dimension
+      \param N7 Entries in 8th-Dimension
+    */
+    KOKKOS_INLINE_FUNCTION
+    explicit constexpr LayoutHilbert3D( size_t N0 = 0, size_t N1 = 0,
+                                        size_t N2 = 0, size_t N3 = 0,
+                                        size_t N4 = 0, size_t N5 = 0,
+                                        size_t N6 = 0, size_t N7 = 0 )
+        : dimension{N0, N1, N2, N3, N4, N5, N6, N7}
+    {
+        // Force N0 and N1 to be 2^n for Square Hilbert Space
+        dimension[0] = (size_t)std::pow(
+            (double)2, (double)ceil( log( (double)N0 ) / log( 2.0 ) ) );
+        dimension[1] = (size_t)std::pow(
+            (double)2, (double)ceil( log( (double)N1 ) / log( 2.0 ) ) );
+        dimension[2] = (size_t)std::pow(
+            (double)2, (double)ceil( log( (double)N2 ) / log( 2.0 ) ) );
+    }
+};
+
+namespace Impl
+{
+//---------------------------------------------------------------------------//
+// HView Offset which implements the index mapping
+//---------------------------------------------------------------------------//
+// ViewOffset struct for LayoutHilbert3D
+template <class Dimension>
+struct ViewOffset<Dimension, Kokkos::LayoutHilbert3D, void>
+{
+
+    using is_mapping_plugin = std::true_type;
+    using is_regular = std::true_type;
+
+    typedef size_t size_type;
+    typedef Dimension dimension_type;
+    typedef Kokkos::LayoutHilbert3D array_layout;
+
+    // Store current dimensions
+    dimension_type m_dim;
+
+    // Store original dimensions of parent view if subview - Store current
+    // dimensions if not subview
+    dimension_type orig_dim;
+
+    // Store offset indices - non-zero if subview
+    dimension_type m_off;
+
+    // Calculate offset if subview. If not subview = 0
+    size_t offset = hilbert3d( m_off.N0, m_off.N1, m_off.N2 );
+
+    // Calculate 3D Hilbert index given an ( x, y, z ) coordinate and the size
+    // of the hilbert space ( n )
+    // Compact 3D Hilbert Indices, using algorithm described in
+    // Compact Hilbert Indices, Dalhousie University Technical Report
+    template <typename I0, typename I1, typename I2>
+    KOKKOS_INLINE_FUNCTION size_t hilbert3d( I0 const &i0, I1 const &i1,
```

Comment:
Let's discuss this implementation. This is nice that this is more general, but can't we treat this as if all views are a "subview" of a space that is `n x n x n` where `n` is the power-of-2 needed to encapsulate the largest dimension of the problem?

Reply:
We can certainly talk about this more, but we could do this. I think the easiest way would be when instantiating the layout, rather than setting each dimension as rounded up to the nearest power-of-2, we take the max dimension and round that up to the nearest power of 2 and then set the the layout to be `n x n x n`. This wouldn't require changing the 3D Hilbert algorithm, but I could look around for a more efficient implementation when all 3 dimensions are equal. Would just come at the expense of requiring extra space that isn't really being used.